nucypher/utilities/versioning.py

```diff
@@ -0,0 +1,166 @@
+
+ nucypher is distributed in the hope that it will be useful,
+ but WITHOUT ANY WARRANTY; without even the implied warranty of
+ MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ GNU Affero General Public License for more details.
+
+ You should have received a copy of the GNU Affero General Public License
+ along with nucypher.  If not, see <https://www.gnu.org/licenses/>.
+"""
+
+
+from abc import abstractmethod, ABC
+from typing import Dict, Tuple, Callable, NamedTuple
+
+
+class Versioned(ABC):
+    """Base class for serializable entities"""
+
+    _PARTS = 2  # bytes
+    _PART_SIZE = 2
+    _BRAND_SIZE = 2
+    _VERSION_SIZE = _PART_SIZE * _PARTS
+    _HEADER_SIZE = _BRAND_SIZE + _VERSION_SIZE
+
+    class InvalidHeader(ValueError):
+        """Raised when an unexpected or invalid bytes header is encountered."""
+
+    class IncompatibleVersion(ValueError):
+        """Raised when attempting to deserialize incompatible bytes"""
+
+    class Empty(ValueError):
+        """Raised when 0 bytes are remaining after parsing the header."""
+
+    @classmethod
+    @abstractmethod
+    def _brand(cls) -> bytes:
+        raise NotImplementedError
+
+    @classmethod
+    @abstractmethod
+    def _version(cls) -> Tuple[int, int]:
+        """tuple(major, minor)"""
+        raise NotImplementedError
+
+    @classmethod
+    def version_string(cls) -> str:
+        major, minor = cls._version()
+        return f'{major}.{minor}'
+
+    #
+    # Serialize
+    #
+
+    def __bytes__(self) -> bytes:
+        return self._header() + self._payload()
+
+    @classmethod
+    def _header(cls) -> bytes:
+        """The entire bytes header to prepend to the instance payload."""
+        major, minor = cls._version()
+        major_bytes = major.to_bytes(cls._PART_SIZE, 'big')
+        minor_bytes = minor.to_bytes(cls._PART_SIZE, 'big')
+        header = cls._brand() + major_bytes + minor_bytes
+        return header
+
+    @abstractmethod
+    def _payload(self) -> bytes:
+        """The unbranded and unversioned bytes-serialized representation of this instance."""
+        raise NotImplementedError
+
+    #
+    # Deserialize
+    #
+
+    @classmethod
+    @abstractmethod
+    def _from_bytes_current(cls, data):
+        """The current deserializer"""
+        raise NotImplementedError
+
+    @classmethod
+    @abstractmethod
+    def _old_version_handlers(cls) -> Dict[Tuple[int, int], Callable]:
+        """Old deserializer callables keyed by version."""
+        raise NotImplementedError
+
+    @classmethod
+    def from_bytes(cls, data: bytes):
+        """"Public deserialization API"""
+        brand, version, payload = cls._parse_header(data)
+        version = cls._get_compatible_handler(data=data, version=version)
+        handlers = cls._deserializers()
+        return handlers[version](payload)
+
+    @classmethod
+    def _get_compatible_handler(cls, data: bytes, version: Tuple[int, int]) -> Tuple[int, int]:
```

Comment:
A small nitpick - a better name would be `_get_compatible_handler_version()`

Reply:
I *just* renamed it to `resolve_version` as you submitted this comment.